server/src/test/java/org/elasticsearch/plugins/PluginsServiceTests.java

```diff
@@ -613,6 +622,66 @@ public void testThrowingConstructor() {
         assertThat(e.getCause().getCause(), hasToString(containsString("test constructor failure")));
     }
 
+    private ClassLoader buildTestProviderPlugin(String name) throws Exception {
+        String pluginClass = Strings.format("""
+            package r;
+            import org.elasticsearch.plugins.spi.TestService;
+            import org.elasticsearch.plugins.ActionPlugin;
+            import org.elasticsearch.plugins.Plugin;
+            public final class FooPlugin extends Plugin implements ActionPlugin, TestService {
```

Comment:
I would separate this out. Due to the current plugin design, a Plugin subclass must be present, but this class should not be required to be the one that implements all service providers. In fact, I think it is important we test them decoupled, since that is the future we are working towards.

Reply:
Sure thing, I was just lazy :). I'll create a separate TestService provider class.

Reply:
Done now. I've split the provider class into separate file.